src/de/impls.rs

```diff
@@ -540,3 +541,222 @@ impl<E: VisitorError, K: FromStr + Ord, V: Jayson<E>> Jayson<E> for BTreeMap<K,
+where
+    E: VisitorError,
+    T: Ord + Jayson<E>,
+{
+    fn begin(out: &mut Option<Self>) -> &mut dyn Visitor<E> {
+        impl<E, T> Visitor<E> for Place<BTreeSet<T>>
+        where
+            T: Jayson<E> + Ord,
+            E: VisitorError,
+        {
+            fn seq(&mut self) -> Result<Box<dyn Seq<E> + '_>, E> {
+                Ok(Box::new(SeqBuilder {
+                    out: &mut self.out,
+                    set: BTreeSet::new(),
+                    value: None,
+                }))
+            }
+        }
+
+        struct SeqBuilder<'a, T: 'a> {
+            out: &'a mut Option<BTreeSet<T>>,
+            set: BTreeSet<T>,
+            value: Option<T>,
+        }
+
+        impl<'a, E, T> Seq<E> for SeqBuilder<'a, T>
+        where
+            T: Jayson<E> + Ord,
+            E: VisitorError,
+        {
+            fn element(&mut self) -> Result<&mut dyn Visitor<E>, E> {
+                if let Some(value) = self.value.take() {
+                    self.set.insert(value);
+                }
+                Ok(Jayson::begin(&mut self.value))
+            }
+
+            fn finish(&mut self) -> Result<(), E> {
+                if let Some(value) = self.value.take() {
+                    self.set.insert(value);
+                }
+                *self.out = Some(mem::take(&mut self.set));
+                Ok(())
+            }
+        }
+
+        Place::new(out)
+    }
+}
+
+impl<T, E> Jayson<E> for HashSet<T>
+where
+    E: VisitorError,
+    T: Hash + Eq + Jayson<E>,
+{
+    fn begin(out: &mut Option<Self>) -> &mut dyn Visitor<E> {
+        impl<E, T> Visitor<E> for Place<HashSet<T>>
+        where
+            T: Jayson<E> + Hash + Eq,
+            E: VisitorError,
+        {
+            fn seq(&mut self) -> Result<Box<dyn Seq<E> + '_>, E> {
+                Ok(Box::new(SeqBuilder {
+                    out: &mut self.out,
+                    set: HashSet::new(),
+                    value: None,
+                }))
+            }
+        }
+
+        struct SeqBuilder<'a, T: 'a> {
+            out: &'a mut Option<HashSet<T>>,
+            set: HashSet<T>,
+            value: Option<T>,
+        }
+
+        impl<'a, E, T> Seq<E> for SeqBuilder<'a, T>
+        where
+            T: Jayson<E> + Hash + Eq,
+            E: VisitorError,
+        {
+            fn element(&mut self) -> Result<&mut dyn Visitor<E>, E> {
+                if let Some(value) = self.value.take() {
+                    self.set.insert(value);
+                }
+                Ok(Jayson::begin(&mut self.value))
+            }
+
+            fn finish(&mut self) -> Result<(), E> {
+                if let Some(value) = self.value.take() {
+                    self.set.insert(value);
+                }
+                *self.out = Some(mem::take(&mut self.set));
+                Ok(())
+            }
+        }
+
+        Place::new(out)
+    }
+}
+
+impl<E: VisitorError> Jayson<E> for serde_json::Value {
```

Comment:
overall, I would prefer this to be behind a compile flag. Let's do that in another pr.